hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/util/NelsonData.java

```diff
@@ -0,0 +1,442 @@
+/*
+ * Copyright 2015-2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.alerts.engine.util;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.apache.commons.math3.stat.descriptive.moment.Mean;
+import org.apache.commons.math3.stat.descriptive.moment.StandardDeviation;
+import org.hawkular.alerts.api.model.condition.NelsonCondition.NelsonRule;
+import org.hawkular.alerts.api.model.data.Data;
+import org.hawkular.alerts.engine.impl.AlertProperties;
+
+/**
+ * For each dataId used in one or more NelsonConditions we have a NelsonData to track NelsonRule violations for that
+ * dataId. The NelsonData is used in the engine (Drools' working memory), it should not be exposed in the API.  The
+ * default sampleSize used to determine mean and standard deviation is 50.  It can be overridden globally via
+ * system property hawkular-alerts.nelson-rules-sample-size or environment variable HAWKULAR_NELSON_RULES_SAMPLE_SIZE.
+ * It can be overridden for a specific dataId via hawkular-alerts.nelson-rules-sample-size.<dataId> or
+ * HAWKULAR_NELSON_RULES_SAMPLE_SIZE_<dataId>.
+ *
+ * @author Jay Shaughnessy
+ * @author Lucas Ponce
+ */
+public class NelsonData {
+    String NELSON_RULES_SAMPLE_SIZE = "hawkular-alerts.nelson-rules-sample-size";
+    String NELSON_RULES_SAMPLE_SIZE_ENV = "HAWKULAR_NELSON_RULES_SAMPLE_SIZE";
+    String NELSON_RULES_SAMPLE_SIZE_DEFAULT = "50";
```

Comment:
Actually, I started with the sampleSize as being part of the NelsonCondition definition.  But that is a problem because *all* NelsonConditions for the same dataId share one NelsonData tracking fact/object.  So it is not possible to have condition-specific sample size (unless we make a significant change and have one NelsonData per Condition. but this is less efficient and at the moment we have no evidence that we need this level of granularity on sample size).  It is obviously not as convenient, but using sysprops we can set sample size per dataId.  I did take a look on the internet about what was a "proper" sample size for mean and standard deviation and in general I saw the size of 20-50 being recommended, so I used 50 as a default.  Collecting 50 can take some time, so I didn't want to go higher.